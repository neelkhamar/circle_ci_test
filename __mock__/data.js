import { ErrorMessages } from "../components/Common/Data/errorMessages";

export const vehicleData = [
  {
    id: 1,
    brand: "Ford",
    model: "Figo",
    year: 2004,
    serie_number: "YDK8348HDI3847",
    catalog_vehicle_type: {
      code: "T_123",
      name: "Ford Figo",
    },
    license_plate: "YDK8348",
    has_freights: false,
    insurer_name: "LIC",
    insurance_policy_number: "Jeevan Bema",
  },
  {
    id: 2,
    brand: "BMW",
    model: "A5",
    year: 2018,
    serie_number: "YDK8348HDI3444",
    catalog_vehicle_type: {
      code: "T_521",
      name: "BMW A5",
    },
    license_plate: "SFK6041",
    has_freights: true,
    insurer_name: "Paisa Bazaar",
    insurance_policy_number: "Jeevan Raksha",
  },
];

export const vehicleInputElements = [
  {
    placeHolder: "Ingrese Placa",
    field: "license_plate",
    value: "YDJ2841",
    required: true,
    errorMessage: ErrorMessages.numberPlate_required,
  },
  {
    placeHolder: "Ingrese Identificacion Vehicular",
    field: "serie_number",
    value: "73248937428729189",
    required: true,
    errorMessage: ErrorMessages.seriel_number_required,
  },
  {
    placeHolder: "Ingrese Marca",
    field: "brand",
    value: "Ford",
    required: true,
    errorMessage: ErrorMessages.brand_required,
  },
  {
    placeHolder: "Ingrese Modelo",
    field: "model",
    value: "SX4",
    required: true,
    errorMessage: ErrorMessages.model_required,
  },
  {
    placeHolder: "Ingrese Ano",
    field: "year",
    value: 2004,
    required: true,
    errorMessage: ErrorMessages.year_required,
  },
  {
    placeHolder: "Ingrese Nombre Aseguradora",
    field: "insurer_name",
    value: "John Doe",
    required: true,
    errorMessage: ErrorMessages.insurer_name_required,
  },
  {
    placeHolder: "Ingrese Numero de Poliza",
    field: "insurance_policy_number",
    value: "6734628767",
    required: true,
    errorMessage: ErrorMessages.insurance_policy_number_required,
  },
];
