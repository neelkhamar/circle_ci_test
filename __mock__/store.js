import configureStore from "redux-mock-store";

export const mockStore = (initialState) => {
  const mockStore = configureStore();

  return mockStore(initialState);
};
