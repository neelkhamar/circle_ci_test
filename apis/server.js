import axios from 'axios';
import alertContainer from '../utils/Alert';
import { logout } from '../redux/user/Action';
import { store } from '../redux/configureStore';
import Router from 'next/router';
import { API_URL } from './constants';

//const url = 'http://localhost:3001/';
const url = API_URL;

let axiosResponse = null;

const apiServer = (options = {}) => {

  const { headers, bufferResponse } = options || {};
  let object = {
    baseURL: url,
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      ...headers
    },
  }
  if(bufferResponse) {
    object["responseType"] = 'arraybuffer';
  }
  axiosResponse = axios.create(object);
  
  axiosResponse.interceptors.response.use(
    (res) => {
      return res;
    },
    (err) => {
      if(err.response.status === 401 && !err.response.config.url.includes("/sign_in")) {
        alertContainer({
          title: 'No autorizado',
          text: err.response.data.errors[0],
          icon: 'error',
          showConfirmButton: false
        });
        store.dispatch(logout());
      }

      if(err.response.status === 402 && !err.response.config.url.includes("/sign_in")) {
        alertContainer({
          title: 'Pago Requerido',
          text: err.response.data.message || err.response.data.errors[0],
          icon: 'warning',
          showConfirmButton: false
        });
        Router.push('/subscription/settings/');
      }
      return Promise.reject(err);
    }
  );
  return axiosResponse;
}



export default apiServer
