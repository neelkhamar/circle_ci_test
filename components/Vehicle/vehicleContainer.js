import {
  DeleteOutlined,
  EditOutlined,
  LoadingOutlined,
} from "@ant-design/icons";
import { Col, Row, Spin, Table } from "antd";
import FeatherIcon from "feather-icons-react";
import { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { deleteVehicle, getVehicles } from "../../requests/carta-porte";
import alertContainer from "../../utils/Alert";
import confirmContainer from "../../utils/SwalConfirm";
import { Button } from "../buttons/buttons";
import { Cards } from "../cards/frame/cards-frame";
import VehicleModal from "../Common/Modals/vehicleModal";
import { PageHeader } from "../page-headers/page-headers";
import { Main } from "./styled";

const antIcon = <LoadingOutlined style={{ fontSize: 40 }} spin />;

const VehicleContainer = () => {
  const [visible, setVisible] = useState(false);
  const [spinner, setSpinner] = useState(false);
  const [data, setData] = useState([]);
  const [selectedVehicle, setSelectedVehicle] = useState(0);

  const [state, setState] = useState({
    selectionType: "checkbox",
    selectedRowKeys: null,
    selectedRows: null,
    values: {},
  });

  useEffect(() => {
    if (selectedVehicle) {
      setVisible(true);
    }
  }, [selectedVehicle]);

  const columnsSort = [
    {
      title: "Marca",
      dataIndex: "brand",
      key: "marca",
      render: (_, item) => {
        return <span data-testid="vehicle-line-item">{item.brand}</span>;
      },
    },
    {
      title: "Modelo",
      dataIndex: "model",
      key: "modelo",
      render: (_, item) => {
        return <span>{item.model}</span>;
      },
    },
    {
      title: "Año",
      dataIndex: "year",
      key: "ano",
      render: (_, item) => {
        return <span>{item.year}</span>;
      },
    },
    {
      title: "Numero de identificacion Vehicular",
      dataIndex: "serie_number",
      key: "numero_de_vehicular",
      render: (_, item) => {
        return <span>{item.serie_number}</span>;
      },
    },
    {
      title: "Tipo de vehiculo",
      dataIndex: "tipo",
      key: "tipo",
      render: (_, item) => {
        return (
          <span>
            {item.catalog_vehicle_type.code +
              " - " +
              item.catalog_vehicle_type.name}
          </span>
        );
      },
    },
    {
      title: "Placa",
      dataIndex: "license_plate",
      key: "placa",
      render: (_, item) => {
        return <span>{item.license_plate}</span>;
      },
    },
    {
      title: "Accion",
      dataIndex: "numero_de_puertas",
      render: (_, item) => {
        return (
          <div className="d-flex" data-testid={`action-container-${item.id}`}>
            <EditOutlined
              data-testid={`edit-icon-${item.id}`}
              className="editIcon mr-3"
              onClick={() => setSelectedVehicle(item.id)}
            />
            {!item.has_freights ? (
              <DeleteOutlined
                data-testid={`delete-icon-${item.id}`}
                className="editIcon ml-3 red-color"
                onClick={() => handleModalOpen(item.id)}
              />
            ) : (
              <></>
            )}
          </div>
        );
      },
      key: "numero_de_puertas",
    },
  ];

  const handleCancel = () => {
    setVisible(false);
    setSelectedVehicle(0);
  };

  const { currentUser } = useSelector((state) => {
    return {
      currentUser: state.auth.currentUser,
    };
  });

  useEffect(() => {
    let mounted = true;
    if (mounted) {
      fetchVehicles();
    }

    return () => (mounted = false);
  }, []);

  const fetchVehicles = () => {
    setSpinner(true);
    getVehicles(
      currentUser.accessToken,
      currentUser.uid,
      currentUser.client
    ).then(
      (response) => {
        setSpinner(false);
        const results = response.data.map((row, index) => {
          row["key"] = index;
          return row;
        });
        setData(results);
      },
      (error) => {
        setSpinner(false);
      }
    );
  };

  const handleOk = () => {
    setVisible(false);
    setSelectedVehicle(0);
    fetchVehicles();
  };

  const handleOpen = (e) => {
    e.preventDefault();
    e.stopPropagation();
    setVisible(true);
  };

  function onChange(pagination, filters, sorter, extra) {
    setState({ ...state, values: { pagination, filters, sorter, extra } });
  }

  const confirmDelete = (id) => {
    setSpinner(true);
    deleteVehicle(
      currentUser.accessToken,
      currentUser.uid,
      currentUser.client,
      id
    ).then(
      (response) => {
        setSpinner(false);
        if (response.status === 204) {
          alertContainer({
            title: "El vehículo fue eliminado con éxito",
            text: "",
            icon: "success",
            showConfirmButton: false,
          });
          fetchVehicles();
        }
      },
      (error) => {
        setSpinner(false);
      }
    );
  };

  const handleModalOpen = (value) => {
    confirmContainer(
      "Eliminar Vehiculos",
      `Estás segura de que quieres eliminar este vehículo?`
    ).then((result) => {
      if (result.isConfirmed) {
        confirmDelete(value);
      }
    });
  };

  return (
    <>
      <PageHeader
        ghost
        title="Vehiculos"
        buttons={[
          <div key="6" className="page-header-actions">
            <Button
              size="small"
              key="4"
              data-testid="create-vehicle"
              type="primary"
              onClick={handleOpen}
            >
              <FeatherIcon icon="plus" size={14} />
              Nuevo
            </Button>
          </div>,
        ]}
      />
      <Main>
        <Row gutter={25}>
          <Col lg={24} xs={24}>
            <Cards headless>
              <div style={{ minHeight: "calc(100vh - 320px)" }}>
                {visible && (
                  <VehicleModal
                    visible={visible}
                    selectedVehicle={selectedVehicle}
                    handleCancel={handleCancel}
                    handleOk={handleOk}
                  />
                )}
                {spinner ? (
                  <div
                    className="text-center pt-4"
                    data-testid="fetch-data-spinner"
                  >
                    <Spin indicator={antIcon} />
                  </div>
                ) : (
                  <Table
                    className="table-responsive"
                    data-testid="vehicle-table"
                    pagination={{
                      defaultPageSize: 10,
                      total: data.length,
                      showTotal: (total, range) =>
                        `${range[0]}-${range[1]} of ${total} items`,
                    }}
                    columns={columnsSort}
                    dataSource={data}
                    onChange={onChange}
                  />
                )}
              </div>
            </Cards>
          </Col>
        </Row>
      </Main>
    </>
  );
};

export default VehicleContainer;
