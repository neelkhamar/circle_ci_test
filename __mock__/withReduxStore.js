import { render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { mockStore } from "./store";

const initialState = {
  auth: {
    currentUser: {
      uid: "joel@gmail.com",
      accessToken: "2374983247",
      client: "78Yuiyaf3849BHJL",
      roles: [],
      userDetails: "",
      certsValidated: false,
    },
  },
};

export const store = mockStore(initialState);

export const renderWithStore = async (Component) => {
  let result = await act(async () =>
    render(<Provider store={store}>{Component}</Provider>)
  );
  return result;
};
