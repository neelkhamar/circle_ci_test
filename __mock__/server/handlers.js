import { rest } from "msw";
import { API_URL } from "../../apis/constants";
import { vehicleData } from "../data";

export const handlers = [
  rest.get(
    `${API_URL}api/v1/vehicles`,
    (req, res, ctx) => {
      return res(ctx.json(vehicleData));
    }
  ),

  rest.get(
    `${API_URL}api/v1/vehicles/types`,
    (req, res, ctx) => {
      return res(
        ctx.json([
          {
            id: 23,
            code: "HTG",
            name: "HTG Transport",
          },
        ])
      );
    }
  ),

  rest.get(
    `${API_URL}api/v1/vehicles/1`,
    (req, res, ctx) => {
      return res(ctx.json(vehicleData[0]));
    }
  ),
];
