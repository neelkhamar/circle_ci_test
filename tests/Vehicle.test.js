import "@testing-library/jest-dom";
import { fireEvent, screen, waitFor } from "@testing-library/react";
import VehicleModal from "../components/Common/Modals/vehicleModal";
import Vehicle from "../pages/moving-note/vehicles";
import { vehicleData, vehicleInputElements } from "../__mock__/data";
import server from "../__mock__/server/server";
import { renderWithStore } from "../__mock__/withReduxStore";

beforeAll(() => server.listen());

describe("Describe Vehicle Component", () => {
  beforeEach(async () => {
    const useRouter = jest.spyOn(require("next/router"), "useRouter");

    useRouter.mockImplementation(() => ({
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: jest.fn(),
      replace: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    }));

    await renderWithStore(<Vehicle />);
  });

  it("Test the title of the screen", async () => {
    const linkElement = screen.getByText(/Vehiculos/i);
    expect(linkElement).toBeInTheDocument();
  });

  it("Check if create vehicle button is displayed on the screen", async () => {
    const linkElement = screen.getByText(/Nuevo/i);
    expect(linkElement).toBeInTheDocument();
  });

  it("Test the fetch vehicle data and display vehicle data", async () => {
    const dataElement = await waitFor(() =>
      screen.getByTestId("vehicle-table")
    );
    expect(dataElement).toBeInTheDocument();

    // Check the row count for vehicle table
    const rowCount = screen.queryAllByTestId("vehicle-line-item");
    expect(rowCount.length).toBe(vehicleData.length);
  });

  it.each(vehicleData)(
    "Test if the delete icon is displayed/hidden properly for the vehicle that are linked to any flete",
    (item) => {
      // Check if delete icon is present or not
      const deleteElement = screen.queryByTestId(`delete-icon-${item.id}`);
      const dataElement = screen.queryByTestId(`action-container-${item.id}`);
      if (!item.has_freights) {
        expect(dataElement).toContainElement(deleteElement);
      } else {
        expect(dataElement).not.toContainElement(deleteElement);
      }
    }
  );

  it.each(vehicleData)("Test if the edit button is displayed in UI", (item) => {
    const dataElement = screen.queryByTestId(`action-container-${item.id}`);
    const editElement = screen.queryByTestId(`edit-icon-${item.id}`);
    expect(dataElement).toContainElement(editElement);
  });

  it("Test the create vehicle button functionality and modal", async () => {
    const linkElement = screen.getByTestId("create-vehicle");
    const baseElement = screen.baseElement;
    let modalElement = screen.queryByTestId("create-vehicle-modal");

    // Modal should be null before clicking on button
    expect(modalElement).toBeNull();
    expect(baseElement).toMatchSnapshot();
    fireEvent.click(linkElement);

    const modalSpinner = screen.queryByTestId("modal-spinner");
    expect(modalSpinner).toBeInTheDocument();

    await renderWithStore(<VehicleModal />);

    modalElement = screen.queryByTestId("create-vehicle-modal");
    expect(modalElement).toBeInTheDocument();
  });

  it("Test the on change functionality for all input boxes in vehicle modal", async () => {
    const linkElement = screen.getByTestId("create-vehicle");
    fireEvent.click(linkElement);

    await renderWithStore(<VehicleModal />);

    vehicleInputElements.forEach((el) => {
      const inputElement = screen.getByPlaceholderText(el.placeHolder);
      fireEvent.change(inputElement, { target: { value: el.value } });

      expect(inputElement).toHaveValue(el.value);
    });
  });

  it("Test the validations for all input boxes in vehicle modal", async () => {
    const linkElement = screen.getByTestId("create-vehicle");
    fireEvent.click(linkElement);

    await renderWithStore(<VehicleModal />);

    fireEvent.click(screen.getByTestId("vehicle-button"));

    vehicleInputElements.forEach(async (el) => {
      await waitFor(() => {
        const errorMessage = screen.queryByText(el.errorMessage);
        expect(errorMessage).toBeInTheDocument();
      });
    });
  });

  it("Test the delete modal dialog box on click of delete icon", async () => {
    const dataElement = await waitFor(() =>
      screen.getByTestId("vehicle-table")
    );
    expect(dataElement).toBeInTheDocument();

    vehicleData.forEach((val) => {
      if (!val.has_freights) {
        const deleteIcon = screen.getByTestId(`delete-icon-${val.id}`);

        expect(deleteIcon).toBeInTheDocument();
        fireEvent.click(deleteIcon);

        expect(document.querySelector(".swal2-container")).toBeInTheDocument();

        const textElement = screen.getByText("Eliminar Vehiculos");
        expect(textElement).toBeInTheDocument();
      }
    });
  });

  it("Test the edit modal on click of edit icon", async () => {
    const dataElement = await waitFor(() =>
      screen.getByTestId("vehicle-table")
    );
    expect(dataElement).toBeInTheDocument();

    const { id } = vehicleData[0];
    const editIcon = screen.getByTestId(`edit-icon-${id}`);

    expect(editIcon).toBeInTheDocument();
    fireEvent.click(editIcon);

    await renderWithStore(<VehicleModal selectedVehicle={id} />);

    const textElement = await waitFor(() =>
      screen.getByText("Actualizar vehÃ­culo")
    );

    // Check if the edit modal is displayed
    expect(textElement).toBeInTheDocument();

    let inputElement = {
      brand: null,
      model: null,
    };

    Object.keys(inputElement).forEach((input) => {
      vehicleInputElements.forEach((formEl) => {
        if (formEl.field === input) {
          inputElement[input] = vehicleData[0][input];
        }
      });
    });

    const { brandElement, modelElement } = await waitFor(() => {
      return {
        brandElement: screen.getByText(inputElement.brand),
        modelElement: screen.getByText(inputElement.model),
      };
    });

    expect(brandElement).toBeInTheDocument();
    expect(modelElement).toBeInTheDocument();
  });
});
