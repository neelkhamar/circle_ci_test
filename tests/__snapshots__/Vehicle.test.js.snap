// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Describe Vehicle Component Test the create vehicle button functionality and modal 1`] = `undefined`;
